RecoHI/HiJetAlgos/plugins/HiPuRhoProducer.cc

```diff
@@ -0,0 +1,615 @@
+          vrms0[vi] = esigma_[it];
+        }
+        rhoExtra_.push_back(vrho0[vi]);
+        nTow_.push_back(vntow[vi]);
+      }
+    } else {
+      emean_[it] = 0.;
+      esigma_[it] = 0.;
+    }
+    LogDebug("PileUpSubtractor") << " ieta: " << it << " Pedestals: " << emean_[it] << "  " << esigma_[it] << "\n";
+  }
+
+  postOrphan_ = false;
+}
+
+void HiPuRhoProducer::subtractPedestal(std::vector<fastjet::PseudoJet>& coll) {
+  LogDebug("PileUpSubtractor") << "The subtractor subtracting pedestals...\n";
+
+  std::vector<fastjet::PseudoJet> newcoll;
+  for (auto& input_object : coll) {
+    int index = input_object.user_index();
+    reco::Candidate const* itow = inputs_[index];
+
+    const CaloTower* ctc = dynamic_cast<const CaloTower*>(itow);
+    int it = ctc->id().ieta();
+
+    double original_Et = itow->et();
+    double etnew = original_Et - (emean_.find(it))->second - (esigma_.find(it))->second;
+    float mScale = etnew / input_object.Et();
+    if (etnew < 0.)
+      mScale = 0.;
+
+    math::XYZTLorentzVectorD towP4(
+        input_object.px() * mScale, input_object.py() * mScale, input_object.pz() * mScale, input_object.e() * mScale);
+
+    input_object.reset_momentum(towP4.px(), towP4.py(), towP4.pz(), towP4.energy());
+    input_object.set_user_index(index);
+
+    if (etnew > 0. && dropZeroTowers_)
+      newcoll.push_back(input_object);
+  }
+
+  if (dropZeroTowers_)
+    coll = newcoll;
+}
+
+void HiPuRhoProducer::calculateOrphanInput(std::vector<fastjet::PseudoJet>& orphanInput) {
+  LogDebug("PileUpSubtractor") << "The subtractor calculating orphan input...\n";
+
+  fjInputs_ = fjOriginalInputs_;
+
+  std::vector<int> jettowers;                       // vector of towers indexed by "user_index"
+  std::vector<std::pair<int, int>> excludedTowers;  // vector of excluded ieta, iphi values
+
+  int32_t nref = 0;
+  for (auto const& pseudojetTMP : fjJets_) {
+    EtaPhi jet_etaphi(pseudojetTMP.eta(), pseudojetTMP.phi());
+    if (nref < nMaxJets_) {
+      jtexngeom[nref] = 0;
+      jtexntow[nref] = 0;
+      jtexpt[nref] = 0;
+      jtpt[nref] = pseudojetTMP.perp();
+      jteta[nref] = pseudojetTMP.eta();
+      jtphi[nref] = pseudojetTMP.phi();
+    }
+
+    if (pseudojetTMP.perp() < puPtMin_)
+      continue;
+
+    for (auto const& im : allgeomid_) {
+      double dr2 =
+          reco::deltaR2(towermap[(DetId)im].first, towermap[(DetId)im].second, jet_etaphi.first, jet_etaphi.second);
+      auto exclude = std::find(excludedTowers.begin(), excludedTowers.end(), std::pair(im.ieta(), im.iphi()));
+      if (dr2 < radiusPU_ * radiusPU_ && exclude == excludedTowers.end() &&
+          (geomtowers_[im.ieta()] - ntowersWithJets_[im.ieta()]) > minimumTowersFraction_ * (geomtowers_[im.ieta()])) {
+        ntowersWithJets_[im.ieta()]++;
+        excludedTowers.emplace_back(im.ieta(), im.iphi());
+
+        if (nref < nMaxJets_)
+          jtexngeom[nref]++;
+      }
+    }
```

Comment:
looking at igprof results (sent earlier), it looks like this loop is rather heavy.
It looks more practical/efficient to :
- loop over ```towermap``` elements instead of ```allgeomid_```, as only ieta and iphi are used
- compute exclude only after ```dr2 <``` check is done
- use an std::map or std::unordered_map to speedup lookup inside the excludedTowers from linear (vector)

Reply:
But `towermap` gives eta and phi coordinates, not ieta and iphi

Reply:
ah, right; looking more carefully, I see that the towermap (a std::map) is used only in conjunction with looping over allgeomid_ (a vector). So, the map functionality of it is not really needed.
For this case it's more appropriate to make it a vector, perhaps even a vector of {eta, phi, ieta, iphi} and use it here instead of allgeomid_.

Reply:
So `towermap` should be `std::vector<std::array<double,4>>`?

Reply:
I meant a struct, like ```struct EtaPhiTower{ int ieta, iphi; float eta, phi;}```
so that access to the values is readable (much better than e.g.  ```tower[i][0]```)